### PR TITLE
Cluster time gossiping

### PR DESCRIPTION
**Pull request by mbroadst**

### CODE REVIEW

File: lib/connection/pool.js [hunk 1]
```diff
@@ -542,6 +544,12 @@ function messageHandler(self) {
           return handleOperationCallback(self, workItem.cb, new MongoError(err));
         }
 
+        // Look for clusterTime, and update it if necessary
+        if (message.documents[0] && message.documents[0].hasOwnProperty('$clusterTime')) {
+          const $clusterTime = message.documents[0].$clusterTime;
```
Comment (kmahar) [inline lib/connection/pool.js]: what is the rationale behind doing this as opposed to just directly assigning `self.topology.clusterTime = message.documents[0].$clusterTime`? I don't see $clusterTime used again

File: lib/connection/pool.js [hunk 2]
```diff
@@ -1134,6 +1142,16 @@ Pool.prototype.write = function(commands, options, cb) {
   // Get the requestId
   operation.requestId = commands[commands.length - 1].requestId;
 
+  if (supportsClusterTime(this.topology) && this.topology.clusterTime) {
+    commands.forEach(command => {
+      if (command instanceof Query) {
+        command.query.$clusterTime = this.topology.clusterTime;
```
Comment (kmahar) [inline lib/connection/pool.js]: thought it was supposed to be a top-level field in the command? why is it different for queries?

File: lib/connection/pool.js [hunk 3]
```diff
@@ -31,6 +31,11 @@ var DESTROYED = 'destroyed';
 
 var _id = 0;
 
+function supportsClusterTime(topology) {
```
Comment (kmahar) [inline lib/connection/pool.js]: should there be a test checking behavior in the case where cluster time isn't supported?

**Comment by kmahar** [comment]
done commenting

File: lib/connection/pool.js (hunk 3 above)
Comment (mbroadst) [reply lib/connection/pool.js]: all existing mock tests use a lower `maxWireVersion`, so essentially "all other tests" are expressing that code path 😄

File: lib/connection/pool.js (hunk 1 above)
Comment (mbroadst) [reply lib/connection/pool.js]: Good catch. The logic for `resolveClusterTime` used to be here, so this was more relevant.

File: lib/connection/pool.js (hunk 2 above)
Comment (mbroadst) [reply lib/connection/pool.js]: Also good catch.  Originally the spec was designed assuming existence of `OP_MSG` support, which the server is guaranteed to support with `maxWireVersion >= 6`.  Unfortunately, we will not necessarily have that done in time so we need to accommodate `OP_QUERY` _and_ `OP_MSG` forms, and the `OP_QUERY` form expects this field to exist within the top-level `query` field.  I've informed @rstam about this.

**Comment by kmahar** [comment]
cool lgtm

**Pull request merged**